File: packages/sdk-kotlin/src/main/kotlin/io/orotp/sdk/internal/HttpExecutor.kt
package io.syrotp.sdk.internal

import io.syrotp.sdk.SyrotpAuthError
import io.syrotp.sdk.SyrotpError
import io.syrotp.sdk.SyrotpNetworkError
import io.syrotp.sdk.SyrotpRateLimitError
import io.syrotp.sdk.SyrotpServerError
import io.syrotp.sdk.SyrotpTimeoutError
import io.syrotp.sdk.SyrotpValidationError
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import okhttp3.Response
import java.io.IOException
import java.net.SocketTimeoutException
import kotlin.math.max

/**
 * Centralizes the retry policy and the HTTP-status-to-exception mapping.
 *
 * Tests can replace [sleeper] to avoid actually sleeping during retries.
 *
 * Retries happen on:
 *  - [SyrotpNetworkError]  (DNS / TLS / connection failures)
 *  - [SyrotpServerError]   (HTTP 5xx)
 *  - [SyrotpRateLimitError] (HTTP 429, honoring `Retry-After`)
 *
 * Never retries on:
 *  - [SyrotpAuthError]
 *  - [SyrotpValidationError]
 *  - [SyrotpTimeoutError]
 *  - any other non-retriable [SyrotpError]
 */
internal class HttpExecutor(
    private val maxRetries: Int,
    private val sleeper: Sleeper = Sleeper { Thread.sleep(it) },
) {

    init {
        require(maxRetries >= 0) { "maxRetries must be >= 0" }
    }

    internal fun interface Sleeper {
        fun sleep(millis: Long)
    }

    internal fun <T> execute(
        transport: () -> Response,
        parse: (Response) -> T,
    ): T {
        var attempt = 0
        while (true) {
            val response: Response = try {
                transport()
            } catch (e: SocketTimeoutException) {
                // Per the contract, timeouts are NOT auto-retried.
                throw SyrotpTimeoutError(message = e.message ?: "request timed out", cause = e)
            } catch (e: IOException) {
                val err = SyrotpNetworkError(
                    message = e.message ?: "network error",
                    cause = e,
                )
                if (attempt < maxRetries) {
                    sleeper.sleep(secondsToMillis(Backoff.seconds(attempt + 1)))
                    attempt++
                    continue
                }
                throw err
            }

            // We never call `parse` on a non-2xx response; map first,
            // throw the typed error, and let the catch decide retry.
            try {
                if (response.isSuccessful) {
                    response.use { return parse(it) }
                }
                response.use { throw errorFromResponse(it) }
            } catch (e: SyrotpError) {
                if (attempt < maxRetries && isRetriable(e)) {
                    val sleepMs = if (e is SyrotpRateLimitError && e.retryAfterSeconds != null) {
                        // Sleep at least Retry-After; jitter MAY add to it.
                        max(
                            secondsToMillis(Backoff.seconds(attempt + 1)),
                            (e.retryAfterSeconds!! * 1000L),
                        )
                    } else {
                        secondsToMillis(Backoff.seconds(attempt + 1))
                    }
                    sleeper.sleep(sleepMs)
                    attempt++
                    continue
                }
                throw e
            }
        }
    }

    private fun isRetriable(e: SyrotpError): Boolean = when (e) {
        is SyrotpNetworkError, is SyrotpServerError, is SyrotpRateLimitError -> true
        else -> false
    }

    private fun secondsToMillis(seconds: Double): Long =
        (seconds * 1000.0).toLong()

    private fun errorFromResponse(response: Response): SyrotpError {
        val status = response.code
        val text = response.body?.string().orEmpty()
        var code = "http_$status"
        var message = "request failed with status $status"
        var requestId: String? = null

        if (text.isNotEmpty()) {
            try {
                val parsed = JSON.parseToJsonElement(text)
                val errObj = (parsed as? JsonObject)?.get("error") as? JsonObject
                if (errObj != null) {
                    (errObj["code"] as? JsonPrimitive)?.contentOrNull?.let { code = it }
                    (errObj["message"] as? JsonPrimitive)?.contentOrNull?.let { message = it }
                    (errObj["request_id"] as? JsonPrimitive)?.contentOrNull?.let { requestId = it }
                }
            } catch (_: Throwable) {
                code = "bad_response"
                message = "non-JSON response (status $status)"
            }
        }

        return when (status) {
            401, 403 -> SyrotpAuthError(code, message, status, requestId)
            400 -> SyrotpValidationError(code, message, status, requestId)
            429 -> SyrotpRateLimitError(
                code = code,
                message = message,
                httpStatus = 429,
                requestId = requestId,
                retryAfterSeconds = parseRetryAfter(response.header("Retry-After")),
            )
            in 500..599 -> SyrotpServerError(code, message, status, requestId)
            else -> SyrotpError(code, message, status, requestId)
        }
    }

    private fun parseRetryAfter(value: String?): Int? {
        if (value.isNullOrBlank()) return null
        val n = value.trim().toIntOrNull() ?: return null
        return if (n < 0) 0 else n
    }

    private companion object {
        private val JSON = Json { ignoreUnknownKeys = true; isLenient = true }
    }
}

File: packages/sdk-swift/Sources/SyrotpSDK/Internal/HttpExecutor.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Centralizes the retry policy and the HTTP-status-to-exception mapping.
///
/// Tests can replace `sleeper` with a no-op closure to skip wall-clock
/// time without spinning real Tasks.
///
/// Retries happen on:
///  - `SyrotpNetworkError`   (DNS / TLS / connection failures)
///  - `SyrotpServerError`    (HTTP 5xx)
///  - `SyrotpRateLimitError` (HTTP 429, honoring `Retry-After`)
///
/// Never retries on:
///  - `SyrotpAuthError`
///  - `SyrotpValidationError`
///  - `SyrotpTimeoutError`
///  - any other non-retriable `SyrotpError`
struct HttpExecutor {
    let maxRetries: Int
    let sleeper: @Sendable (TimeInterval) async -> Void

    init(
        maxRetries: Int,
        sleeper: @escaping @Sendable (TimeInterval) async -> Void = { seconds in
            let nanos = UInt64(max(0, seconds) * 1_000_000_000)
            try? await Task.sleep(nanoseconds: nanos)
        }
    ) {
        precondition(maxRetries >= 0, "maxRetries must be >= 0")
        self.maxRetries = maxRetries
        self.sleeper = sleeper
    }

    /// Run `transport()` up to `maxRetries + 1` times. On retriable
    /// failures, sleep with backoff and try again.
    ///
    /// `parse` is called only on 2xx; the executor itself maps non-2xx
    /// responses to typed errors and decides whether to retry.
    func execute<T>(
        transport: @Sendable () async throws -> (Data, HTTPURLResponse),
        parse: @Sendable (Data, HTTPURLResponse) throws -> T
    ) async throws -> T {
        var attempt = 0
        while true {
            let data: Data
            let response: HTTPURLResponse
            do {
                (data, response) = try await transport()
            } catch let urlError as URLError where urlError.code == .timedOut {
                // Per the contract: SDK does NOT auto-retry timeouts —
                // the caller's deadline already expired.
                throw SyrotpTimeoutError(
                    message: urlError.localizedDescription,
                    underlying: urlError
                )
            } catch let urlError as URLError where urlError.code == .cancelled {
                // Caller cancelled their Task — bubble it as
                // CancellationError, NOT a timeout.
                throw CancellationError()
            } catch {
                let err = SyrotpNetworkError(
                    message: error.localizedDescription,
                    underlying: error
                )
                if attempt < maxRetries {
                    await sleeper(Backoff.seconds(for: attempt + 1))
                    attempt += 1
                    continue
                }
                throw err
            }

            // 2xx: parse and return.
            if (200..<300).contains(response.statusCode) {
                return try parse(data, response)
            }

            // Non-2xx: map to typed error. Retry if retriable and budget remains.
            let typed = errorFromResponse(data: data, response: response)
            if attempt < maxRetries, isRetriable(typed) {
                let sleepSeconds: TimeInterval
                if let rate = typed as? SyrotpRateLimitError, let retryAfter = rate.retryAfterSeconds {
                    // Sleep at LEAST Retry-After; jitter MAY make it longer.
                    sleepSeconds = max(Backoff.seconds(for: attempt + 1), TimeInterval(retryAfter))
                } else {
                    sleepSeconds = Backoff.seconds(for: attempt + 1)
                }
                await sleeper(sleepSeconds)
                attempt += 1
                continue
            }
            throw typed
        }
    }

    private func isRetriable(_ error: SyrotpError) -> Bool {
        switch error {
        case is SyrotpNetworkError, is SyrotpServerError, is SyrotpRateLimitError:
            return true
        default:
            return false
        }
    }

    private func errorFromResponse(data: Data, response: HTTPURLResponse) -> SyrotpError {
        let status = response.statusCode
        var code = "http_\(status)"
        var message = "request failed with status \(status)"
        var requestId: String? = nil

        if !data.isEmpty {
            if let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let errObj = parsed["error"] as? [String: Any] {
                if let c = errObj["code"] as? String { code = c }
                if let m = errObj["message"] as? String { message = m }
                if let r = errObj["request_id"] as? String { requestId = r }
            } else if (try? JSONSerialization.jsonObject(with: data)) == nil {
                code = "bad_response"
                message = "non-JSON response (status \(status))"
            }
        }

        switch status {
        case 401, 403:
            return SyrotpAuthError(code: code, message: message, httpStatus: status, requestId: requestId)
        case 400:
            return SyrotpValidationError(code: code, message: message, httpStatus: status, requestId: requestId)
        case 429:
            let retryAfter = parseRetryAfter(response.value(forHTTPHeaderField: "Retry-After"))
            return SyrotpRateLimitError(
                code: code,
                message: message,
                httpStatus: 429,
                requestId: requestId,
                retryAfterSeconds: retryAfter
            )
        case 500...599:
            return SyrotpServerError(code: code, message: message, httpStatus: status, requestId: requestId)
        default:
            return SyrotpError(code: code, message: message, httpStatus: status, requestId: requestId)
        }
    }

    private func parseRetryAfter(_ value: String?) -> Int? {
        guard let v = value?.trimmingCharacters(in: .whitespaces), !v.isEmpty else { return nil }
        guard let n = Int(v) else { return nil }
        return max(0, n)
    }
}
